Sync documentation (#2799)
msupply-foundation/open-msupply: server/service/src/sync/README.md

> # Sync - Syncrhonisation

Reviewer: ```suggestion
# Sync - Synchronisation
```

Open mSupply is often used in environments with intermittent and low bandwidth internet and therefore needs to be accessible in offline mode. 

> A mechanism that allows for this to happen is called **Sync** (or **Syncrhonisation**).

Reviewer: ```suggestion
A mechanism that allows for this to happen is called **Sync** (or **Synchronisation**).
```

> Open mSupply is often used in environments with intermittent and low bandwidth internet and therefore needs to be accessible in offline mode. 
> A mechanism that allows for this to happen is called **Sync** (or **Synchronisation**).

Reviewer: :nit Just a suggestion

```suggestion
Open mSupply is often used in environments with intermittent and/or low bandwidth internet access and therefore needs to be accessible in an offline mode. 
A mechanism that allows for this to happen is called **Synchronisation**, or **Sync** for short.
```

## Glossary

> `site`: An instance of Open mSupply app with one or more active store

Reviewer: ```suggestion
`site`: An instance of Open mSupply with one or more active stores
```

`remote site`: This is the same as a site, but we quite often refer to it as ‘remote site’ to emphasise that it’s not always accessible via the internet, and even though it’s part of the same system, it’s not always directly connected to the system.

> `central server`/`configuration server`: The main site of mSupply system, this is where central data is configured, including credentials for remote sites, users and active store associations with sites. This is also the server that ‘drives’ synchronisation, all remote sites will synchronise with the central server, which in turn can forward records to other sites (for example, in case of transfer).

Reviewer: maybe we should just pick one term and use that 🤔 I wonder if it's potentially confusing using the two

Author: I would prefer that tbh, what i was trying to do with the 'alternatives' was to help new devs, that haven't heard these concepts before, to show them that it's the same thing. ¯\_(ツ)_/¯, i think the first term is the primary term

Reviewer: yeah, that's good. if someone hears 'configuration server' then they might know that it means the same as 'central server'. Was thinking more generally that we need to make an effort to settle on one term for things and avoid confusion

## Record Types

Records are treated differently by central server and remote sites based on data types, for example remote records belong to only one store and are only editable on a site where this store is active, whereas central data is only editable on a central server, the former usually syncs from remote site to central server whereas the later will only travel from central server to remote site.

`remote record/data`: Can only be owned and edited by one store, on a site where this store is active, this data typically travels from remote site where it’s created and edited, to central server, which in turn may decide to forward it to another remote site (in case of transfers). Shipment, requisition, stock_line are examples of remote record.
 
`central record/data`: This record is only editable on central server, and only travels in one direction, from central server to all remote sites. All central data will sync to all remote sites. Items and units are examples of central records.

> `central-remote record/data`: Similar in editability to central data, but will only sync to remote sites where these records are relevant. For example, master_list or name_store_join.

Reviewer: ```suggestion
`central-remote record/data`: Similar in edibility to central data, but will only sync to remote sites where these records are relevant. For example, master_list or name_store_join.
```

Author: @mark-prins, hmm i guess editability is not technically a word ? But i've seen it used in places ¯\_(ツ)_/¯ 
sorry I am struggling to ingest the suggestion : )

Reviewer 2: Might want to rephrase it altogether since edibility isn't the correct word for this either 😆 

Though... You can eat data (:

Reviewer: confused face: `edibility` is a word. It's just spelt edibility and not editability for some reason 🤷

Reviewer 2: `Definitions of edibility. the property of being fit to eat. synonyms: edibleness. types: digestibility, digestibleness. the property of being easy to digest.`

Reviewer: oops

Reviewer: ok, `These records are editable in the same way as central data, but it will only sync to remote sites where these records are relevant. For example, master_list or name_store_join.`

Author: Thanks: 202442a2e00142e979213c6e4941a33e2f8ec7bb

`transfer record/data`: This is remote record but for a store that is not active on current site, this record is visible but not editable. For example, request requisition in response site.

`shared-remote record/data`: Editable on any site and sync to all sites. For example barcodes

`patient record/data`: Editable and accessible on any site where the patient is visible. Visibility in this case is determined by name store join. Prescribers, patients and patient documents are examples of patient record.

> ## Phases/Stages of Sychronisation:

Reviewer: ```suggestion
## Phases/Stages of Synchronisation:
```

`initialisation`: After the site is configured, it will first need to be initialised. This is where all relevant central and remote records are queued and sent to a remote site. Remote site is not ‘usable’ until initialisation stage is completed and all records are received

> `operational`: Once initialised the site will be in operational synchronisation phase, any changes to remote data will be sent to central server, and any changes to central data will travel on remote site from central server.

Reviewer: Is this what you mean? I wasn't sure about the 'travel on'
```suggestion
`operational`: Once initialised the site will be in an operational synchronisation phase, any changes to remote data will be sent to central server, and any changes to central data will travel to the remote site from the central server.
```

Author: Yeah ! travel to, thanks

> We also consider central server as `backup service` for remote sites, in case of data loss at remote site, it can be re-initialised (usually by clearing app data, and switching site back to initialisation phase). Central server protects multiple instances of mSupply syncing for the same site by recording hardware id during each sync API request, if hardware id does not match previously stored hardware id an error will be returned.

Reviewer: should it be 'open mSupply'?
```suggestion
We also consider central server as `backup service` for remote sites, in case of data loss at remote site, it can be re-initialised (usually by clearing app data, and switching site back to initialisation phase). Central server protects multiple instances of Open mSupply syncing for the same site by recording a device identifier ( hardware id ) during each sync API request, if the hardware id does not match the previously stored hardware id then an error will be returned.
```

Author: > should it be 'open mSupply'?

It's actually for both, but I guess Open mSupply makes it less confusing, since we haven't mentioned mSupply anywhere in the docs as remote site, thanks

## How does a remote site work ?

> A worked thread called [SynchroniserDriver](https://github.com/msupply-foundation/open-msupply/blob/bc83acbb3cd51fe3375ac01135c6eb880a793936/server/service/src/sync/synchroniser_driver.rs#L11) will run synchroniser every [interval_seconds](https://github.com/msupply-foundation/open-msupply/blob/bc83acbb3cd51fe3375ac01135c6eb880a793936/server/configuration/example.yaml#L20). [SynchroniserDriver](https://github.com/msupply-foundation/open-msupply/blob/bc83acbb3cd51fe3375ac01135c6eb880a793936/server/service/src/sync/synchroniser.rs#L88) will also listen to a message that can trigger sync manually at any time.

Reviewer: ```suggestion
A worker thread called [SynchroniserDriver](https://github.com/msupply-foundation/open-msupply/blob/bc83acbb3cd51fe3375ac01135c6eb880a793936/server/service/src/sync/synchroniser_driver.rs#L11) will run synchroniser every [interval_seconds](https://github.com/msupply-foundation/open-msupply/blob/bc83acbb3cd51fe3375ac01135c6eb880a793936/server/configuration/example.yaml#L20). [SynchroniserDriver](https://github.com/msupply-foundation/open-msupply/blob/bc83acbb3cd51fe3375ac01135c6eb880a793936/server/service/src/sync/synchroniser.rs#L88) will also listen to a message that can trigger sync manually at any time.
```

## Integrations

Translations will return an array of boxed [Upsert or Delete](https://github.com/msupply-foundation/open-msupply/blob/bc83acbb3cd51fe3375ac01135c6eb880a793936/server/repository/src/lib.rs#L49) traits, which are stored in IntegrationOperation enum. These traits are implemented by every sync data type, in the repository layer. 
To ensure that the whole record in its entirety (invoice and all it’s lines) is available after synchronisation, we integrate everything in one big transaction. However this does slow down the initialisation phase, and thus transaction is skipped during initialisation, this is safe because the app is not available during the initialisation stage.

## Example of Adding translation

Please see below commit for an example of adding a new table and relative translations (it’s only an example and assumes that central server can already handle this record).

[Add asset table](https://github.com/msupply-foundation/open-msupply/commit/dfa5b8e4a4d0a906b0c19f3ee00e194344834508)

[Add translations](https://github.com/msupply-foundation/open-msupply/commit/a9b3d90a597f9670687f0fc26a5632310fe6b4c9)

## Diagrams

> From [TMF internal docs](https://app.diagrams.net/#G1vVZQF_o0JiABBs1jpI0TneFfRonVqySD)

Reviewer: this is nit picky! there's a spelling error in the linked diagram `syncrhoniser`

Author: Ouch, i'll update it in the downstream PR, since diagram has changed, thanks

Author: 62d85d5bf8c6c37ad41a7476f9d10d4b7a16eec5, also moved diagram to code base, with Draw.io extension it's really cool
